File: 86-partition-list/86-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        head1, head2 = ListNode(0), ListNode(0)
        p1, p2 = head1, head2
        
        cur = head
        
        while(cur):
            nxt = cur.next
            if cur.val < x:
                p1.next = cur
                cur.next = None
                p1 = p1.next
            else:
                p2.next = cur
                cur.next = None
                p2 = p2.next                
            cur = nxt
        p1.next = head2.next
        return head1.next
                

File: 86-partition-list/86-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode dumb = new ListNode(0);
        ListNode small = dumb;
        ListNode noSmallDumb = new ListNode(0);
        ListNode noSmall = noSmallDumb;
        ListNode  cur = head;
        ListNode nxt;
        while(cur != null){
            nxt = cur.next;
            if(cur.val < x){
                small.next = cur;
                cur.next = null;
                small = small.next;
            }else{
                noSmall.next = cur;
                cur.next = null;
                noSmall = noSmall.next;
            }
            cur = nxt;
        }
        small.next = noSmallDumb.next;
        return dumb.next;
    }
}